### ISSUE TITLE
Critical point loss SimCC

### ISSUE THREAD

**Issue by Egrt** [ultralytics/ultralytics#4243]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

![image](https://github.com/ultralytics/ultralytics/assets/47631482/792aa5c7-5d91-4992-a722-3b5460f0a907)
The latest key point detection model adopts SimCC's loss function, and how to modify the detection header and decoding to adapt to better performance SimCC losses？

```
class Pose(Detect):

    def __init__(self, nc=80, kpt_shape=(17, 3), ch=()):
        """Initialize YOLO network with default parameters and Convolutional Layers."""
        super().__init__(nc, ch)
        self.kpt_shape = kpt_shape  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
        self.nk = kpt_shape[0] * kpt_shape[1]  # number of keypoints total
        self.detect = Detect.forward

        c4 = max(ch[0] // 4, self.nk)
        self.cv4 = nn.ModuleList(nn.Sequential(Conv(x, c4, 3), Conv(c4, c4, 3), nn.Conv2d(c4, self.nk, 1)) for x in ch)

    def forward(self, x):
        """Perform forward pass through YOLO model and return predictions."""
        bs = x[0].shape[0]  # batch size
        kpt = torch.cat([self.cv4[i](x[i]).view(bs, self.nk, -1) for i in range(self.nl)], -1)  # (bs, 17*3, h*w)
        x = self.detect(self, x)
        if self.training:
            return x, kpt
        pred_kpt = self.kpts_decode(bs, kpt)
        return torch.cat([x, pred_kpt], 1) if self.export else (torch.cat([x[0], pred_kpt], 1), (x[1], kpt))

    def kpts_decode(self, bs, kpts):
        """Decodes keypoints."""
        ndim = self.kpt_shape[1]
        if self.export:  # required for TFLite export to avoid 'PLACEHOLDER_FOR_GREATER_OP_CODES' bug
            y = kpts.view(bs, *self.kpt_shape, -1)
            a = (y[:, :, :2] * 2.0 + (self.anchors - 0.5)) * self.strides
            if ndim == 3:
                a = torch.cat((a, y[:, :, 2:3].sigmoid()), 2)
            return a.view(bs, self.nk, -1)
        else:
            y = kpts.clone()
            if ndim == 3:
                y[:, 2::3].sigmoid_()  # inplace sigmoid
            y[:, 0::ndim] = (y[:, 0::ndim] * 2.0 + (self.anchors[0] - 0.5)) * self.strides
            y[:, 1::ndim] = (y[:, 1::ndim] * 2.0 + (self.anchors[1] - 0.5)) * self.strides
            return y
```

### Use case

Compared to the current loss function, SimCC can significantly improve the performance of key point detection.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Egrt hi there,

Great to see your interest in improving the YOLOv8 keypoint detection model with the SimCC loss function! Indeed, using different loss functions can indeed significantly impact the performance of a model.

In order to modify the detection header and decoding to adapt to SimCC losses, you would need to modify the loss computation specifically for the keypoints within the Pose class you posted. However, please note that you should also consider the impact of introducing a new loss function on other parts of the model such as the object detection. Careful integration is required to ensure that improving keypoint detection wouldn't adversely affect other features of YOLOv8.

It's key to thoroughly test these modifications to ensure stability and performance. Also, take note that while the SimCC loss function might improve model performance on a specific dataset, we're aiming for a generalized solution fit for a variety of datasets.

Your willingness to help by submitting a PR is also much appreciated. Collaboration and contributions like this help in ongoing enhancement of YOLOv8 and broader YOLO community.

Thank you!
